### PR TITLE
Fix memory leaks and possible invalid reads

### PR DESCRIPTION
**Pull request by JohannesLorenz**
Should be easy to review.

I hope is that there were no assumptions that some memory was planned to exist until program termination (i.e. the leaks are wanted).

### CODE REVIEW

**Comment by JohannesLorenz** [comment]
I'm converting this to a draft for now. It works, but I think the `if(impl)` query in the DTOR is over-careful?

File: src/cpp/ports.cpp [hunk 1]
```diff
@@ -508,8 +508,10 @@ Ports::Ports(std::initializer_list<Port> l)
 
 Ports::~Ports()
 {
-    delete []impl->enump;
-    delete impl;
+    if(impl) {
```
Comment (fundamental) [inline src/cpp/ports.cpp:511]: I would just add a Port_Matcher::~Port_Matcher(void) method and then you can just have "delete impl". Let C++ handle the implicit if(x) get_rid_of(x) logic for you.

**Comment by fundamental** [comment]
Both changes look reasonable enough to me. Ports are at the global scope, so not much of a difference there, the automation ports though are likely leaking data every time a new master instance is loaded. Your call if you want to address the one comment or not as it's mostly a cosmetic thing than anything. Thanks for the fix :+1:

File: src/cpp/ports.cpp [hunk 2]
```diff
@@ -835,11 +840,10 @@ char *Ports::collapsePath(char *p)
 void Ports::refreshMagic()
 {
     delete impl;
-    impl = new Port_Matcher;
+    impl = new Port_Matcher(ports.size());
```
Comment (JohannesLorenz) [inline src/cpp/ports.cpp:843]: I moved allocating `impl->enump` here, because I think it should never be dependent on `generate_minimal_hash`.

**Comment by JohannesLorenz** [comment]
@fundamental Re-worked your comment. Can you please check review the new commit?

One of the travis CI jobs is hanging, but it's probably just a force push issue. The final commit passed.

**Comment by fundamental** [comment]
Looks good here. Thanks for identifying the old bugs :)

**Pull request merged**